tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php

```diff
@@ -55,6 +62,191 @@ public function setProductTaxRulesGroup(string $taxRulesGroupReference, string $
+        $product = $this->getProductByReference($productReference);
+        $productCategories = $product->getCategories();
+
+        $belongsToDefaultCategory = false;
+        foreach ($productCategories as $categoryId) {
+            if ((int) $categoryId === $defaultCategoryId) {
+                $belongsToDefaultCategory = true;
+
+                break;
+            }
+        }
+
+        if ((int) $product->id_category_default !== $defaultCategoryId || !$belongsToDefaultCategory) {
+            throw new RuntimeException('Default category is not assigned to product');
+        }
+    }
+
+    /**
+     * @Then product :productReference type should be :productType
+     *
+     * @param string $productReference
+     * @param string $productTypeName
+     */
+    public function assertProductType(string $productReference, string $productTypeName)
+    {
+        $product = $this->getProductByReference($productReference);
+        $productTypeValue = $this->getProductTypeValueByName($productTypeName);
+
+        $isVirtual = $product->is_virtual;
+        $isPack = Pack::isPack($product->id);
+        $isStandard = !Pack::isPack($product->id) && !$product->is_virtual;
+        $isCombination = !empty($product->getAttributeCombinations());
+
+        if ($isVirtual && $productTypeValue === ProductType::TYPE_VIRTUAL) {
+            return;
+        }
+
+        if ($isPack && $productTypeValue === ProductType::TYPE_PACK) {
+            return;
+        }
+
+        if ($isStandard && $productTypeValue === ProductType::TYPE_STANDARD) {
+            return;
+        }
+
+        if ($isCombination && $productTypeValue === ProductType::TYPE_COMBINATION) {
+            return;
+        }
+
+        throw new RuntimeException(sprintf('Product type is not as expected. Expected %s', $productTypeName));
+    }
+
+    /**
+     * @Then I should get error that product name is invalid
+     */
+    public function assertLastErrorIsReturnProductDisabled()
+    {
+        $this->assertLastErrorIs(
+            ProductConstraintException::class,
+            ProductConstraintException::INVALID_NAME
+        );
+    }
+
+    /**
+     * @Then I should get error that product type is invalid
+     */
+    public function assertLastErrorIsInvalidTypeConstraint()
+    {
+        $this->assertLastErrorIs(
+            ProductConstraintException::class,
+            ProductConstraintException::INVALID_PRODUCT_TYPE
+        );
+    }
+
+    /**
+     * @param string $typeName
+     *
+     * @return int
+     */
+    private function getProductTypeValueByName(string $typeName): int
+    {
+        $typeValueByName = [
+            'standard' => ProductType::TYPE_STANDARD,
+            'pack' => ProductType::TYPE_PACK,
+            'virtual' => ProductType::TYPE_VIRTUAL,
+            'combination' => ProductType::TYPE_COMBINATION,
+            // undefined value to check the real behavior when type is wrong
+            'undefined' => 500,
```

Comment:
It's weird to introduce a specific "undefined" use case, I understand why you need it to check invalid values though
I'd suggest only managing valid values, and for other you can change the return the provided string as is:

`return $typeValueByName[$typeName] ?? $typeName` or `return $typeValueByName[$typeName] ?? -1` to match the expected `int` type